api/db-migrations/000013_add_default_traffic_rule.down.sql

```diff
@@ -0,0 +1,65 @@
+WITH default_traffic_rule_routes AS (
+    SELECT id , default_traffic_rule -> 'routes' AS routes FROM router_versions
+),
+route_tbl AS (
+    SELECT id, jsonb_array_elements(routes) AS route FROM router_versions
+),
+original_route_tbl AS (
+    SELECT id, route -> 'id' AS route_id FROM route_tbl
+),
+-- all routes for each router version
+all_routes AS (
+    SELECT
+        id, jsonb_agg(DISTINCT route_id) routes
+    FROM original_route_tbl
+    GROUP BY id
+),
+dangling_routes AS (
```

Comment:
For the down query the `all_routes` is `all_rule_routes` right? Meaning, we have to only take the routes that appear in the rules. Not all routes :O

Reply:
`all_routes` include all the routes for each router version. When we run the `EXCEPT` clause, it only selects from the `default_traffic_rule_routes` and exclude those where all routes match that particular row's default traffic rule's routes. This is just to capture the case where all routes have been used in `traffic_rules` column, and that there are no  dangling routes to be removed.

Reply:
> When we run the EXCEPT clause, it only selects from the default_traffic_rule_routes and exclude those where all routes match that particular row's default traffic rule's routes

I'm still confused. The `default_traffic_rule_routes` will always be a subset of `all_routes` which means `dangling_routes` will always be empty, no?

And this part below can be done earlier:
```
-- WHERE clause filters away rows with no traffic rules
    WHERE t.routes IS NOT NULL
```

The goal for the down query, wrt to updating the traffic_rules column in:
- If len(default_traffic_rule_routes) == len(all_routes) for a given id
- Else remove the default_traffic_rule_routes from each rule ? Now, on hindsight, is this even correct? What if the user created the `default_traffic_rule` with their own combination of routes?

This makes me think it's better if we don't touch the `traffic_rules` column at all during the down query - no harm keeping the routes in the rules instead of dangling, right?

Reply:
>The default_traffic_rule_routes will always be a subset of all_routes which means dangling_routes will always be empty, no?

Yes, its always a subset, but what we're excluding here is a subset but **without equality**, which points to only router versions where all routes have been used in `traffic_rules` and we're setting all routes for `default_traffic_rule`.

>If len(default_traffic_rule_routes) == len(all_routes) for a given id

Yes, `If len(default_traffic_rule_routes) == len(all_routes) for a given id`, we won't do anything.

>What if the user created the default_traffic_rule with their own combination of routes?

Then this query is wrong. We're not capturing that case, cause there's no way to. I see this rollback used mainly in the event of rolling out only.

Reply:
> Yes, its always a subset, but what we're excluding here is a subset but without equality, which points to only router versions where all routes have been used in traffic_rules and we're setting all routes for default_traffic_rule.

Not sure I follow. If `default_traffic_rule_routes` had [{1, "route-1"}] and `all_routes` had [{1, "route-1"}, {1, "route-2"}], what is the result of `dangling_routes` ?

Reply:
>Not sure I follow. If default_traffic_rule_routes had [{1, "route-1"}] and all_routes had [{1, "route-1"}, {1, "route-2"}], what is the result of dangling_routes

`dangling_routes` will be empty in this case.

Reply:
So what's the point of this query? If `default_traffic_rule_routes` is a subset of `all_routes`, this EXCEPT query will always result in an empty table, no? It's like set operation A - B.

Reply:
> Sorry, my typo above, it should be with equality - dangling_routes will be empty in this case.

Hmm.. Can you add an example of what you mean by "with equality" and "without" ?

Reply:
>So what's the point of this query? If default_traffic_rule_routes is a subset of all_routes, this EXCEPT query will always result in an empty table, no? It's like set operation A - B.

It'll be able to track rules with dangling routes and revert in cases like this:
`default_traffic_rule_routes`
```
 id  |       routes
------+---------------------
 1 | ["route-1", "route-2"]
 ```

`all_routes`
```
 id  |       routes
------+---------------------
 1 | ["route-1", "route-2", "route-3"]
 ```

`dangling_routes`
```
 id  |       routes
------+---------------------
 1 | ["route-1", "route-2"]
 ```

Then, `route-1` and `route-2` will be removed from all traffic_rules' routes.

Reply:
>Hmm.. Can you add an example of what you mean by "with equality" and "without" ?

**with equality**
`default_traffic_rule_routes`
```
 id  |       routes
------+---------------------
 1 | ["route-1", "route-2", "route-3"]
 ```

`all_routes`
```
 id  |       routes
------+---------------------
 1 | ["route-1", "route-2", "route-3"]
 ```

**without equality**
`default_traffic_rule_routes`
```
 id  |       routes
------+---------------------
 1 | ["route-1", "route-2"]
 ```

`all_routes`
```
 id  |       routes
------+---------------------
 1 | ["route-1", "route-2", "route-3"]
 ```

Reply:
What about:

`default_traffic_rule_routes`
```
 id  |       routes
------+---------------------
 1 | ["route-1", "route-2", "route-3"]
```

and 

`all_routes`
```
 id  |       routes
------+---------------------
 1 | ["route-2", "route-1", "route-3"]
```
?

Isn't it simpler just to compare the lengths (`len(default_traffic_rule_routes) == len(all_routes)` ) ?

Reply:
I've updated the query, lmk if it makes sense 🙏🏻